flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -27,9 +27,31 @@ void TableColumn(TableCreationContext context) :
         UniqueKey(context.uniqueKeysList)
     |
         ComputedColumn(context)
+    |
+        Watermark(context)
     )
 }
 
+void Watermark(TableCreationContext context) :
+{
+    SqlIdentifier columnName;
+    SqlParserPos pos;
+    SqlNode watermarkStrategy;
+}
+{
+    <WATERMARK> {pos = getPos();} <FOR>
+    columnName = CompoundIdentifier()
+    <AS>
+    watermarkStrategy = Expression(ExprContext.ACCEPT_NON_QUERY) {
+        if (context.watermark != null) {
+            throw SqlUtil.newContextException(pos,
+                ParserResource.RESOURCE.multipleWatermarksUnsupported());
```

Comment:
Why not using normal Exception here? It looks really complex(by introducing new a Resource concept) to just throw an exception..

Reply:
There is nothing to do with Expression syntax here. This is in order to give a better exception message (with proper position) when multiple watermarks are defined.